crypto/fips/self_test_kats.c

```diff
@@ -0,0 +1,363 @@
+    return NULL;
+}
+
+/* Utility function to setup a DSA EVP_KEY */
+static EVP_PKEY *dsa_set_pkey(BIGNUM *p, BIGNUM *q, BIGNUM *g,
+                              BIGNUM *pub, BIGNUM *priv)
+{
+    EVP_PKEY *pkey = NULL;
+    DSA *dsa = NULL;
+
+    pkey = EVP_PKEY_new();
+    if (pkey == NULL)
+        goto err;
+
+    dsa  = DSA_new();
+    if (dsa == NULL
+            || !DSA_set0_pqg(dsa, p, q, g)
+            || !DSA_set0_key(dsa, pub, priv))
+        goto err;
+
+    EVP_PKEY_assign(pkey, EVP_PKEY_DSA, dsa);
+    return pkey;
+
+err:
+    DSA_free(dsa);
+    EVP_PKEY_free(pkey);
+    return NULL;
+}
+
+/* Utility function to find keydata by id and convert to a BIGNUM */
+static BIGNUM *keydata2bn(ST_KEYDATA *list, int id)
+{
+    int i;
+
+    for (i = 0; list[i].key.data != NULL; ++i) {
+        if (list[i].id == id) {
+            return BN_bin2bn(list[i].key.data, list[i].key.len, NULL);
+        }
+    }
+    return NULL;
+}
+
+/* Utility function to find keydata by id and convert to a string */
+static const char *keydata2str(ST_KEYDATA *list, int id)
+{
+    int i;
+
+    for (i = 0; list[i].key.data != NULL; ++i) {
+        if (list[i].id == id) {
+            return (const char *)list[i].key.data;
+        }
+    }
+    return NULL;
+}
+
+/* Utility function to load a EVP_KEY from binary key data */
+static EVP_PKEY *bin2pkey(const EVP_PKEY_METHOD *meth, ST_KEYDATA *data)
+{
+    if (meth == &rsa_pkey_meth)
+        return rsa_set_pkey(keydata2bn(data, RSA_P),
+                            keydata2bn(data, RSA_Q),
+                            keydata2bn(data, RSA_N),
+                            keydata2bn(data, RSA_E),
+                            keydata2bn(data, RSA_D),
+                            keydata2bn(data, RSA_DMP1),
+                            keydata2bn(data, RSA_DMQ1),
+                            keydata2bn(data, RSA_IQMP));
+    else if (meth == &ec_pkey_meth)
+        return ec_set_pkey(keydata2str(data, EC_CURVE),
+                           keydata2bn(data, EC_X),
+                           keydata2bn(data, EC_Y),
+                           keydata2bn(data, EC_D));
+    else if (meth == &dsa_pkey_meth)
+        return dsa_set_pkey(keydata2bn(data, DSA_P),
+                            keydata2bn(data, DSA_Q),
+                            keydata2bn(data, DSA_G),
+                            keydata2bn(data, DSA_PUB),
+                            keydata2bn(data, DSA_PRIV));
+    else
+        return NULL;
+}
+
+/* Test a single KAT for digest algorithm */
+static int self_test_digest(ST_DIGEST *t, ST_EVENT *event)
+{
+    int ret = 0;
+    EVP_MD_CTX *ctx = NULL;
+    const EVP_MD *md = NULL;
+    unsigned char out[EVP_MAX_MD_SIZE];
+
+    SELF_TEST_EVENT_onbegin(event, SELF_TEST_TYPE_KAT_DIGEST, t->desc);
+
+    md = EVP_get_digestbyname(t->md_name);
+    if (md == NULL)
+        return 0;
+
+    ctx = EVP_MD_CTX_new();
+    if (ctx == NULL)
+        return 0;
+
+    if (!md->init(ctx))
```

Comment:
Why are you calling the "init" function directly rather than using EVP_DigestInit_ex()? This is likely to break with the provider changes I'm putting in, in this area.

Reply:
It needs to stay inside the FIPS boundary -- that rules out calling EVP doesn't it?

If the EVP level calls are available inside, I agree that this should use them.  I guess it depends on what your changes will be.  Changing over ought to be straightforward.

Reply:
The current design assumes we will have copies of EVP code inside the FIPS boundary. From the "FIPS Module Boundary" section of the design doc:

> Note that within the FIPS module we are using EVP concepts (EVP_MD_CTX, EVP_PKEY_CTX, etc) in order to implement this. These are copies of the EVP implementation found in libcrypto. The FIPS module is not linked against libcrypto. This is to ensure that the complete operation stays within the boundary of the FIPS module without calling code that is external to it.

Reply:
If only life was that simple :)
So the self tests are supposed to run BEFORE any of the services are registered (i.e- the providers method tables have not been registered yet). This means that the only mechanism available is to call lower level API's (maybe via the method table functions).
So I am not really sure what to call yet for most of the KATS.